backend/src/Tabulators/IRV.ts

```diff
@@ -0,0 +1,241 @@
+            let topRemainingRankIndex: number = 0
+            let isOverVote = false
+            remainingCandidates.forEach(candidate => {
+                // loop over remaining candidates
+                // find top (non-zero) rank, if duplicate top ranks mark as overvote
+                if (ballot[candidate.index] !== 0) {
+                    // candidate has rank
+
+                    if (topRemainingRank == 0) {
+                        // set initial top rank
+                        topRemainingRankIndex = candidate.index
+                        topRemainingRank = ballot[candidate.index]
+                        isOverVote = false
+                    }
+                    else if (ballot[candidate.index] < ballot[topRemainingRankIndex]) {
+                        // set new top rank
+                        topRemainingRankIndex = candidate.index
+                        topRemainingRank = ballot[candidate.index]
+                        isOverVote = false
+                    }
+                    else if (ballot[candidate.index] === ballot[topRemainingRankIndex]) {
+                        // multiple top ranks, mark as over vote
+                        isOverVote = true
+                    }
+                }
+            })
+            if (isOverVote) {
+                // ballot is overvote
+                overVoteCount += 1
+            }
+            if (topRemainingRank === 0 || isOverVote) {
+                // ballot is exhausted
+                exhaustedVoteCount += 1
+                exhaustedVotes.push(...activeVotes.splice(i, 1))
+            }
+            else {
+                // give vote to top rank candidate
+                roundVoteCounts[topRemainingRankIndex] += 1
+            }
+        }
+        results.voteCounts.push(roundVoteCounts)
+        results.exhaustedVoteCounts.push(exhaustedVoteCount)
+        results.overVoteCounts.push(overVoteCount)
+        let voteThreshold = activeVotes.length / 2
+        
+        // get max number of votes
+        let remainingCandidatesIndexes = remainingCandidates.map(c => c.index)
+        let maxVotes = Math.max(...roundVoteCounts.filter((c, i) => remainingCandidatesIndexes.includes(i)))
+        if (maxVotes >= voteThreshold) {
+            // candidate meets the threshold
+            // get index of winning candidate
+            let winningCandidateIndex = roundVoteCounts.indexOf(maxVotes)
+            // add winner, remove from remaining candidates
+            results.elected.push(summaryData.candidates[winningCandidateIndex])
+            remainingCandidates = remainingCandidates.filter(c => !results.elected.includes(c))
+            // TODO: Some STV stuff
+        }
+        else {
+            // find candidate with least votes and remove from remaining candidates
+            let lastPlaceCandidateIndex = roundVoteCounts.indexOf(Math.min(...roundVoteCounts.filter((c, i) => remainingCandidatesIndexes.includes(i))))
```

Comment:
What if multiple candidates are tied for least votes? I'm not sure what the standard practice is there, should we eliminate all tied non-first place candidates?

Reply:
According to [this](https://legislature.maine.gov/statutes/21-A/title21-Asec723-A.html) its determined by lottery. We could hook up the random candidate selector but I'm not all that concerned about getting this 100% correct at the moment as its just for demos.